drivers/vcnl40x0/vcnl40x0.c

```diff
@@ -0,0 +1,172 @@
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+#define DEV_I2C             (dev->params.i2c_dev)
+#define DEV_ADDR            (dev->params.i2c_addr)
+
+/* Internal functions */
+static int _set_command_reg(const vcnl40x0_t *dev, uint8_t reg)
+{
+    if (i2c_write_reg(DEV_I2C, DEV_ADDR, VCNL40X0_REG_COMMAND, reg, 0) != 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int _get_command_reg(const vcnl40x0_t *dev, uint8_t *reg)
+{
+    if (i2c_read_reg(DEV_I2C, DEV_ADDR, VCNL40X0_REG_COMMAND, reg, 0) != 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+/*---------------------------------------------------------------------------*
+ *                          VCNL40X0 Core API                                *
+ *---------------------------------------------------------------------------*/
+
+int vcnl40x0_init(vcnl40x0_t *dev, const vcnl40x0_params_t *params)
+{
+    dev->params = *params;
+
+    /* Acquire exclusive access */
+    i2c_acquire(DEV_I2C);
+
+    /* Check sensor ID */
+    uint8_t checkid;
+    i2c_read_reg(DEV_I2C, DEV_ADDR, VCNL40X0_REG_PRODUCT_ID,
+                  &checkid, 0);
+    if ((checkid & VCNL40X0_PRODUCT_MASK_PRODUCT_ID) != VCNL40X0_PRODUCT_ID) {
+        DEBUG("[vcnl40x0] Error: wrong device ID '0x%02x', expected '0x%02x'\n",
+              checkid, VCNL40X0_PRODUCT_ID);
+        i2c_release(DEV_I2C);
+        return -VCNL40X0_ERR_NODEV;
+    }
+
+    /* LED current cannot be above 20 */
+    if (dev->params.led_current > 20) {
+        dev->params.led_current = 20;
+    }
+
+    if (i2c_write_reg(DEV_I2C, DEV_ADDR, VCNL40X0_REG_PROXIMITY_CURRENT,
+                                        dev->params.led_current, 0) != 0) {
+        DEBUG("[vcnl40x0] Error: failed to set proximity current\n");
+        i2c_release(DEV_I2C);
+        return -VCNL40X0_ERR_I2C;
+    }
+
+    if (_set_command_reg(dev, VCNL40X0_COMMAND_ALL_DISABLE) < 0) {
+        DEBUG("[vcnl40x0] Error: failed to set disable all commands\n");
+        i2c_release(DEV_I2C);
+        return -VCNL40X0_ERR_I2C;
+    }
+
+    if (i2c_write_reg(DEV_I2C, DEV_ADDR, VCNL40X0_REG_PROXIMITY_RATE,
+                      dev->params.proximity_rate, 0) != 0) {
+        DEBUG("[vcnl40x0] Error: failed to set proximity rate\n");
+        i2c_release(DEV_I2C);
+        return -VCNL40X0_ERR_I2C;
+    }
+
+    if (i2c_write_reg(DEV_I2C, DEV_ADDR, VCNL40X0_REG_AMBIENT_PARAMETER,
+                      dev->params.ambient_rate |
+                      VCNL40X0_AMBIENT_PARA_AUTO_OFFSET_ENABLE |
+                      dev->params.ambient_avg, 0) != 0) {
+        DEBUG("[vcnl40x0] Error: failed to set ambient light rate\n");
+        i2c_release(DEV_I2C);
+        return -VCNL40X0_ERR_I2C;
+    }
+
+    /* Release I2C device */
+    i2c_release(DEV_I2C);
+
+    DEBUG("[vcnl40x0] info: vcnl40x0 sensor initialized with success\n");
+
+    return VCNL40X0_OK;
+}
+
+uint16_t vcnl40x0_read_proximity(const vcnl40x0_t *dev)
+{
```

Comment:
> How are errors handled in that way, for example I2C communication errors, e.g., in case of timeouts or arbitration losses?

Simple, they are not, as you can see. This is also what is in the guide and not everyone agrees with this. I think the argument was memory consumption save.
I would be in favor of using return codes and passing references. If you request this, I can change it.

Reply:
I just asked because I saw your discussion with @smlng about this topic. And of course I've already seen it in the guide and I'm really not happy with it. Not returning an error code is not fault tolerant. Even if the application does not want to evaluate every error, it should at least have the chance to detect errors.
Since your functions are conform with drivers guide, it is Ok for me an no need to change.